File: heuristic.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <utility>
#include <string>
#include <filesystem>
#include "Job.h"
#include "utilities.h"

using namespace std;

/* 
    Calculate the minimum due date (MDD) for a given task.
    Args:
        - C is the completion time of the previous task
        - p is the processing time of the current task
        - d is the due date of the current task
    Returns:
        MDD = max{C + p, d}
*/
long long mdd(long long processed, Job task) {
    return max(processed + task.ptime, task.dDate);
}


/*
    Partition unscheduled jobs into two sets for a machine based on due dates.
    Args:
        - unscheduled_jobs is a list of unscheduled jobs
        - processed is the sum of processing times of the jobs that have already been scheduled on the machine
        - u1 contains jobs that cannot be completed by their due date
        - u2 contains jobs that can be completed by their due date

*/
void partition_jobs(vector<Job> unscheduled_jobs, long long processed, vector<Job>& u1, vector<Job>& u2) {
    for (int i = 0; i < unscheduled_jobs.size(); i++)
        if (processed + unscheduled_jobs[i].ptime > unscheduled_jobs[i].dDate)
            u1.push_back(unscheduled_jobs[i]);
        else
            u2.push_back(unscheduled_jobs[i]);
}

/*
    For a machine, find the job with the minimum MDD value from the unscheduled jobs.
    Args:
        - unscheduled_jobs is a list of unscheduled jobs
        - processed is the sum of processing times of the jobs that have already been scheduled on the machine
    Returns:
        - min_mdd_job is the job with the minimum MDD value
        - min_mdd is the minimum MDD value

*/
pair<Job, long long> find_min_mdd_job(vector<Job> unscheduled_jobs, long long processed) {
    /*
        Find the job with the minimum MDD value from the unscheduled jobs.
    */
    long long min_mdd = LLONG_MAX; 
    Job min_mdd_job;

    for (int i = 0; i < unscheduled_jobs.size(); i++) {
        int mdd2 = mdd(processed, unscheduled_jobs[i]);
        if (mdd2 < min_mdd) {
            min_mdd = mdd2;
            min_mdd_job = unscheduled_jobs[i];
        }
    }

    return make_pair(min_mdd_job, min_mdd);
}

/*
    Implementation of the MDD algorithm for the TT (Total Tardiness) problem with multiple identical machines.
    Args:
        - jobs is a list of all jobs
        - numMachines is the number of machines
    Returns:
        - schedules is a list of schedules for each machine

*/
vector<vector<Job>> mddScheduling(vector<Job> jobs, int numMachines) {
    vector<long long> completion_times(numMachines, 0);  // Initialize completion times for all machines

    vector<vector<Job>> schedules(numMachines);  // Initialize schedules for all machines

    // While there are unscheduled jobs
    while (!jobs.empty()) {
        long long min_delta = LLONG_MAX;
        int min_delta_machine = -1;
        Job min_delta_job;

        for (int machine = 0; machine < numMachines; machine++) {
            vector<Job> u1, u2; // Initialize u1 and u2
            partition_jobs(jobs, completion_times[machine], u1, u2);

            // find the minimum processing time and due date for u1 and u2, respectively
            long long min_process_time = LLONG_MAX;
            long long min_due_date = LLONG_MAX;

            for (Job job : u1)
                min_process_time = min(min_process_time, job.ptime);

            for (Job job : u2)
                min_due_date = min(min_due_date, job.dDate);

            vector<Job> gamma_lambda;
            for (Job job : u1) 
                if (job.ptime == min_process_time)
                    gamma_lambda.push_back(job);

            for (Job job : u2) 
                if (job.dDate == min_due_date) 
                    gamma_lambda.push_back(job);
            
            // find the job with the minimum MDD value from gamma and lambda
            pair<Job, long long> min_mdd_job = find_min_mdd_job(gamma_lambda, completion_times[machine]);

            if (min_mdd_job.second < min_delta) {
                min_delta = min_mdd_job.second;
                min_delta_machine = machine;
                min_delta_job = min_mdd_job.first;
            }
            
        }

        // schedule the job with the minimum MDD value on the machine with the minimum delta
        schedules[min_delta_machine].push_back(min_delta_job);
        // update the completion time of the machine
        completion_times[min_delta_machine] += min_delta_job.ptime;

        // erase the job from the list of unscheduled jobs with its id
        for (int i = 0; i < jobs.size(); i++) {
            if (jobs[i].id == min_delta_job.id) {
                jobs.erase(jobs.begin() + i);
                break;
            }
        }
    }

    return schedules;
} 

int main() {
    // for each file of benchmarks (n=20, m=2), read the file and run the MDD algorithm
    string path = "benchmarks/n=20/m=2/";
    
    for (const auto & entry : filesystem::directory_iterator(path)) {
        // store the information of the file
        int n, m;
        vector<Job> jobs;
        int flag = read_fileinput(entry.path(), jobs, n, m);

        if (flag == -1) {
            cout << "Error reading the file " << entry.path() << endl;
            return -1;
        }

        // Run the MDD algorithm for the TT problem with n jobs and m machines
        vector<vector<Job>> schedules = mddScheduling(jobs, m);

    
        // print the schedule
        for (int i = 0; i < m; i++) {
            cout << "Machine " << i + 1 << ": ";
            for (Job job : schedules[i]) {
                cout << job.id << " ";
            }
            cout << endl;
        }
        
        // calculate total tardiness
        long long total_tardiness = 0;
        for (int i = 0; i < m; i++) {
            long long completion_time = 0;
            for (Job job : schedules[i]) {
                completion_time += job.ptime;
                long long tardiness = max(0LL, completion_time - job.dDate);
                total_tardiness += tardiness;
            }
        } 

        cout << "Total Tardiness: " << total_tardiness << endl;

    }
    
    return 0;
}

File: heuristic.py
import sys
import os

# Define a simple Job class
class Job:
    def __init__(self, idjob, process_time, due_date):
        self.idjob = idjob
        self.process_time = process_time
        self.due_date = due_date

    def __repr__(self):
        return f"Job({self.process_time}, {self.due_date})"
    

def mdd(processed, task):
    """
    Calculate the minimum due date (MDD) for a given task.
    """
    return max(processed + task.process_time, task.due_date)

def partition_jobs(unscheduled_jobs, partition_time):
    """
    Partition unscheduled jobs into two sets based on due dates.
    """
    u1 = [job for job in unscheduled_jobs if partition_time + job.process_time > job.due_date]
    u2 = [job for job in unscheduled_jobs if partition_time + job.process_time <= job.due_date]
    return u1, u2

def find_min_mdd_job(unscheduled_jobs, processed):
    """
    Find the job with the minimum MDD value from the unscheduled jobs.
    """
    min_mdd_job = None
    min_mdd = sys.maxsize

    for job in unscheduled_jobs:
        mdd2 = mdd(processed, job)
        if mdd2 < min_mdd:
            min_mdd = mdd2
            min_mdd_job = job

    return min_mdd_job, min_mdd

def mdd_tt_multiple_machines(jobs, num_machines):
    """
    Implement the MDD algorithm for the TT (Total Tardiness) problem with multiple machines.
    """
    completion_times = [0] * num_machines  # Initialize completion times for all machines
    schedule = [[] for _ in range(num_machines)]  # Initialize schedules for all machines

    while jobs:
        min_delta = sys.maxsize
        min_delta_machine = None
        min_delta_job = None

        for machine in range(num_machines):
            unscheduled_jobs = jobs.copy()
            partition_time = completion_times[machine]
            u1, u2 = partition_jobs(unscheduled_jobs, partition_time)

            # in u1 find the minimum processing time 
            # in u2 find the minimum due date

            min_process_time = sys.maxsize
            min_due_date = sys.maxsize

            for job in u1:
                min_process_time = min(min_process_time, job.process_time)
            
            for job in u2:
                min_due_date = min(min_due_date, job.due_date)

            gamma = [job for job in u1 if job.process_time == min_process_time]
            lambda_ = [job for job in u2 if job.due_date == min_due_date]

            job, delta = find_min_mdd_job(gamma + lambda_, completion_times[machine])

            if delta < min_delta:
                min_delta = delta
                min_delta_machine = machine
                min_delta_job = job

        schedule[min_delta_machine].append(min_delta_job)
        completion_times[min_delta_machine] += min_delta_job.process_time
        jobs.remove(min_delta_job)

    return schedule

if __name__ == "__main__":
    # resultados del benchmakrs 
    results = []
    filenames = []

    # Iterate over each file in the benchmarks/m=2 folder
    for filename in os.listdir("benchmarks/m=2"):
        filepath = os.path.join("benchmarks/m=2", filename)

        filenames.append(filename)

        # Define jobs by input from file
        jobs = []

        savedNM = False  # is n and m saved?
        n = 0
        m = 0

        with open(filepath) as f:
            for line in f:
                # ignore lines that start with '#'
                if line.startswith("#"):
                    continue

                # if n and m have not been saved, save them
                if not savedNM:
                    n, m = map(int, line.strip().split())
                    savedNM = True
                    continue

                idjob, process_time, due_date = map(int, line.strip().split())
                jobs.append(Job(idjob, process_time, due_date))

        # Run the MDD algorithm for the TT problem with 2 machines
        schedule = mdd_tt_multiple_machines(jobs, 2)

        # Print the schedule
        for i, machine_schedule in enumerate(schedule):
            print(f"Machine {i + 1}: {machine_schedule}")

        # Calculate total tardiness
        total_tardiness = 0
        for machine_schedule in schedule:
            completion_time = 0
            for job in machine_schedule:
                completion_time += job.process_time
                tardiness = max(0, completion_time - job.due_date)
                total_tardiness += tardiness

        print(f"Total tardiness: {total_tardiness}")
        results.append(total_tardiness)
    
    for i, filename in enumerate(filenames):
        print(f"File: {filename}, Total Tardiness: {results[i]}")